File: shared/models/whats_new.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from shared.definitions.whats_new import NewTone


class NewItem(BaseModel):
    id: str
    kind: str
    at: datetime
    value: str
    detail: str | None = None
    tone: str = NewTone.NEW.value
    status: int | None = None
    title: str | None = None
    tech: list[str] = Field(default_factory=list)
    ips: list[str] = Field(default_factory=list)
    source: str | None = None
    source_label: str | None = None
    screenshot_path: str | None = None
    severity: str | None = None
    is_kev: bool = False
    sensitive: bool = False
    asset_type: str | None = None
    query: str | None = None
    target_id: uuid.UUID | None = None
    target_value: str | None = None
    target_type: str | None = None
    scan_id: uuid.UUID | None = None
    watch_id: uuid.UUID | None = None
    host_id: uuid.UUID | None = None
    scope_id: uuid.UUID | None = None
    platform: str | None = None
    handle: str | None = None
    program_name: str | None = None
    program_url: str | None = None
    importable: bool = False
    target_exists: bool | None = None
    scanned: bool | None = None
    muted: bool = False


class NewSection(BaseModel):
    kind: str
    total: int
    items: list[NewItem] = Field(default_factory=list)


class NewSubject(BaseModel):
    kind: str
    id: str | None = None
    label: str
    target_id: uuid.UUID | None = None
    target_value: str | None = None
    target_type: str | None = None
    platform: str | None = None
    handle: str | None = None
    watched: bool = False
    watch_id: uuid.UUID | None = None


class NewGroup(BaseModel):
    id: str
    subject: NewSubject
    at: datetime
    counts: dict[str, int] = Field(default_factory=dict)
    sections: list[NewSection] = Field(default_factory=list)
    scan_id: uuid.UUID | None = None
    scan_started_at: datetime | None = None
    scan_status: str | None = None
    previous_scan_id: uuid.UUID | None = None
    retired: int = 0
    run_label: str | None = None


class NewDay(BaseModel):
    date: str
    counts: dict[str, int] = Field(default_factory=dict)


class NewFeed(BaseModel):
    since: datetime
    until: datetime | None = None
    basis: str
    marked_at: datetime | None = None
    window: str | None = None
    counts: dict[str, int] = Field(default_factory=dict)
    facts: dict[str, dict[str, int]] = Field(default_factory=dict)
    daily: list[NewDay] = Field(default_factory=list)
    groups: list[NewGroup] = Field(default_factory=list)
    truncated: bool = False
    first_runs: int = 0
    visual: int = 0
    new_checks: int = 0


class VisualPair(BaseModel):
    id: str
    host: str
    at: datetime
    distance: int
    before_path: str
    after_path: str
    before_status: int | None = None
    after_status: int | None = None
    before_title: str | None = None
    after_title: str | None = None
    before_tech: list[str] = Field(default_factory=list)
    after_tech: list[str] = Field(default_factory=list)
    before_server: str | None = None
    after_server: str | None = None
    moved: list[str] = Field(default_factory=list)
    silent: bool = False
    target_id: uuid.UUID
    target_value: str
    target_type: str
    scan_id: uuid.UUID
    previous_scan_id: uuid.UUID
    query: str


class VisualFeed(BaseModel):
    since: datetime
    until: datetime | None = None
    basis: str
    window: str | None = None
    total: int = 0
    silent: int = 0
    pairs: list[VisualPair] = Field(default_factory=list)
    truncated: bool = False


class NewMark(BaseModel):
    marked_at: datetime


class NewUnseen(BaseModel):
    count: int
    since: datetime | None = None

File: frontend/src/lib/types/whats-new.ts
export interface NewItem {
	id: string;
	kind: string;
	at: string;
	value: string;
	detail: string | null;
	tone: string;
	status: number | null;
	title: string | null;
	tech: string[];
	ips: string[];
	source: string | null;
	source_label: string | null;
	screenshot_path: string | null;
	severity: string | null;
	is_kev: boolean;
	sensitive: boolean;
	asset_type: string | null;
	query: string | null;
	target_id: string | null;
	target_value: string | null;
	target_type: string | null;
	scan_id: string | null;
	watch_id: string | null;
	host_id: string | null;
	scope_id: string | null;
	platform: string | null;
	handle: string | null;
	program_name: string | null;
	program_url: string | null;
	importable: boolean;
	target_exists: boolean | null;
	scanned: boolean | null;
	muted: boolean;
}

export interface NewSection {
	kind: string;
	total: number;
	items: NewItem[];
}

export interface NewSubject {
	kind: string;
	id: string | null;
	label: string;
	target_id: string | null;
	target_value: string | null;
	target_type: string | null;
	platform: string | null;
	handle: string | null;
	watched: boolean;
	watch_id: string | null;
}

export interface NewGroup {
	id: string;
	subject: NewSubject;
	at: string;
	counts: Record<string, number>;
	sections: NewSection[];
	scan_id: string | null;
	scan_started_at: string | null;
	scan_status: string | null;
	previous_scan_id: string | null;
	retired: number;
}

export interface NewDay {
	date: string;
	counts: Record<string, number>;
}

export interface NewFeed {
	since: string;
	until: string | null;
	basis: string;
	marked_at: string | null;
	window: string | null;
	counts: Record<string, number>;
	facts: Record<string, Record<string, number>>;
	daily: NewDay[];
	groups: NewGroup[];
	truncated: boolean;
	first_runs: number;
	visual: number;
}

export interface VisualPair {
	id: string;
	host: string;
	at: string;
	distance: number;
	before_path: string;
	after_path: string;
	before_status: number | null;
	after_status: number | null;
	before_title: string | null;
	after_title: string | null;
	before_tech: string[];
	after_tech: string[];
	before_server: string | null;
	after_server: string | null;
	moved: string[];
	silent: boolean;
	target_id: string;
	target_value: string;
	target_type: string;
	scan_id: string;
	previous_scan_id: string;
	query: string;
}

export interface VisualFeed {
	since: string;
	until: string | null;
	basis: string;
	window: string | null;
	total: number;
	silent: number;
	pairs: VisualPair[];
	truncated: boolean;
}

export interface VisualParams {
	since?: string;
	window?: string;
	day?: string;
	day_to?: string;
	target_id?: string;
	q?: string;
}

export interface NewMark {
	marked_at: string;
}

export interface NewUnseen {
	count: number;
	since: string | null;
}

export interface NewFeedParams {
	since?: string;
	window?: string;
	day?: string;
	day_to?: string;
	target_id?: string;
	platform?: string;
	handle?: string;
	kinds?: string;
	ring?: string;
	q?: string;
}
